File: core-vm/java/java/lang/Object.java
/**************************************************************************
* Copyright (c) 2001 by Punch Telematix. All rights reserved.             *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix nor the names of                 *
*    other contributors may be used to endorse or promote products        *
*    derived from this software without specific prior written permission.*
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX OR OTHER CONTRIBUTORS BE LIABLE       *
* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR            *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF    *
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR         *
* BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,   *
* WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE    *
* OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN  *
* IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                           *
**************************************************************************/


/*
** $Id: Object.java,v 1.1.1.1 2004/07/12 14:07:46 cvs Exp $
*/

package java.lang;

/**
 ** The mother of all classes.
 **
 ** Note: this class is initialized "by hand" before the VM is fully
 ** initialized.  Consequently it must not have a static initializer.
 ** (It can have static variables, and even constant initial values
 ** for those variables, but nothing fancier and certainly no static{}
 ** clause.)
 */
public class Object {
  
  /**
   ** Get the class of which this object is an instance.
   */
  public final native Class getClass();

  /**
   ** Default String representation: overriden by many classes.
   */
  public String toString() {
    return this.getClass().getName()+"@"+Integer.toHexString(this.hashCode());
  }

  /**
   ** The (default) object hashcode: for transient objects this is simply the 
   ** address of the instance in memory, for persistent objects it is the
   ** address at which the instance was first created (maybe in a previous
   ** activation of the VM).
   */
  public native int hashCode();

  /**
   ** Default object equality (is identity).
   */
  public native boolean equals(Object obj);

  /**
   ** Default clone() method performs a "shallow" cloning of the instance.
   */
  protected native Object clone() 
    throws CloneNotSupportedException;

  /**
   ** Wait forever.
   */
  public final void wait() throws IllegalMonitorStateException, InterruptedException {
    wait(0, 0);
  }

  /**
   ** Wait for x ms, 0 ns.
   */
  public final void wait(long millis) throws IllegalMonitorStateException, InterruptedException {
    wait(millis, 0);
  }

  /**
   ** Wait on a monitor.
   */
  public final native void wait(long millis, int nanos) 
    throws IllegalMonitorStateException, InterruptedException;

  /**
   ** Notify (at most) one thread that is waiting on a monitor.
   */
  public final native void notify()
    throws IllegalMonitorStateException;

  /**
   ** Notify all threads (if any) that are waiting on a monitor.
   */
  public final native void notifyAll()
    throws IllegalMonitorStateException;

  /**
   ** The default finalizer does nothing.
   */
  protected void finalize()
    throws Throwable {}

  /**
   ** Private method which is called by the garbage collector (from native
   ** code) to invoke the finalizer of an object.
   */
  private static void do_finalize(Object o) {
    try {
      o.finalize();
    }
    catch (Throwable t) {
    }
  }

}  

File: core-vm/src/native/java/lang/Object.c
/**************************************************************************
* Parts copyright (c) 2001, 2002 by Punch Telematix.                      *
* All rights reserved.                                                    *
* Parts copyright (c) 2004, 2005, 2006, 2008, 2011 by Chris Gray,         *
* /k/ Embedded Java Solutions. All rights reserved.                       *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

#include <string.h>

#include "arrays.h"
#include "checks.h"
#include "clazz.h"
#include "exception.h"
#include "loading.h"
#include "locks.h"
#include "Math.h"
#include "methods.h"
#include "oswald.h"
#include "wstrings.h"


static inline w_int objectHashCode(w_instance instance) {

  return (w_int)instance;

}


w_instance Object_clone(JNIEnv *env, w_instance thisObject) {
  w_thread thread = JNIEnv2w_thread(env);
  w_clazz  clazz = instance2clazz(thisObject);
  w_instance theClone = NULL;

  threadMustBeSafe(thread);
  if (mustBeInitialized(clazz) == CLASS_LOADING_FAILED) {

    return NULL;

  }

  if(clazz->dims) {
    theClone =  cloneArray(thread, thisObject);
    woempa(1, "object is an array: new instance @ %p\n",theClone);
  } else if(implementsInterface(clazz,clazzCloneable)) {

    w_int instanceSize = clazz->instanceSize;
    w_int i;

    enterUnsafeRegion(thread);
    theClone = allocInstance(thread, clazz);

    if (theClone) {
      woempa(1, "cloneable object is of class %k, size is %d words, new instance @ %p\n", clazz, instanceSize,theClone);
      for(i=0;i<instanceSize;++i)
      
        theClone[i] = thisObject[i];
    }
    enterSafeRegion(thread);

  }
  else {
    throwException(thread, clazzCloneNotSupportedException, NULL);
  }

  return theClone;
}

w_instance 
Object_getClass (
  JNIEnv *env, w_instance thisObject
) {
  w_clazz clazz = instance2clazz(thisObject);

  woempa(1, "Getting class instance of %k.\n", instance2clazz(thisObject));

  return clazz2Class(clazz);

}

w_boolean Object_equals(JNIEnv *env, w_instance thisObject, w_instance thatObject) {

  if (thisObject == thatObject) {
    return WONKA_TRUE;
  }

  return WONKA_FALSE;
  
}

w_int Object_hashCode(JNIEnv *env, w_instance thisObject) {
  w_int hashcode;

  woempa(1, "Getting hash code of %j.\n", thisObject);
  
  hashcode = objectHashCode(thisObject);
  woempa(1, "Instance %p hash code is 0x%08x\n", thisObject, hashcode);

  return hashcode;
}

void Object_wait(JNIEnv *env, w_instance thisObject, w_long millis, w_int nanos) {
  w_thread thread = JNIEnv2w_thread(env);
  x_sleep  sleep_ticks = 0;
 
  if(millis < 0 || nanos < 0 || nanos >= 1000000) {
    throwException(thread,clazzIllegalArgumentException,NULL);
  }

 
  if (isNotSet(instance2flags(thisObject), O_HAS_LOCK)) {
    throwException(thread, clazzIllegalMonitorStateException, "lock has no owner");

    return;
  }

  if (testForInterrupt(thread)) {
    return;
  }

  /*
   * The combination millis==0, nanos==0 means "wait forever".
   * But any nonzero number of nanos would mean "wait for a very short time" ...
   */
  if (millis == 0) {
    if (nanos == 0) {
      sleep_ticks = x_eternal;
    }
    else {
      sleep_ticks = 1;
    }
  }

  /* [CG 20110126, see comment in Thread_wait0 dated 20070509]
  ** This isn't ideal, because the time taken to go around the loop is
  ** additional to the wait time, so we could build up a cumulative
  ** excess of expectancy. However I don't think it's worth it to try
  ** to read the system clock and perform arithmetic on it.
  ** All of this because x_sleep is 32 bits instead of 64 ...
  */
  while (millis > THREE_WEEKS_MILLIS) {
    waitMonitor(thisObject, THREE_WEEKS_TICKS);
    millis -= THREE_WEEKS_MILLIS;

    if (testForInterrupt(thread)) {
      return;
    }
  }

  if (!sleep_ticks) {
    sleep_ticks = x_millis2ticks((w_size)millis);
  }
  if (sleep_ticks) {
    waitMonitor(thisObject, sleep_ticks);
  }

  testForInterrupt(thread);
}

void Object_notify(JNIEnv *env, w_instance thisObject) {

  woempa(1, "Env %p; Notify single thread waiting on instance of %k.\n", env, instance2clazz(thisObject));

  if (isNotSet(instance2flags(thisObject), O_HAS_LOCK)) {
    throwException(JNIEnv2w_thread(env), clazzIllegalMonitorStateException, "not owner");

    return;
  }

  notifyMonitor(thisObject, NOTIFY);
  
}

void Object_notifyAll(JNIEnv *env, w_instance thisObject) {

  woempa(1, "Env %p; Notify all threads waiting on instance of %k.\n", env, instance2clazz(thisObject));

  if (isNotSet(instance2flags(thisObject), O_HAS_LOCK)) {
    throwException(JNIEnv2w_thread(env), clazzIllegalMonitorStateException, "not owner");

    return;
  }

  notifyMonitor(thisObject, NOTIFY_ALL);
  
}

